orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pollers/EphemeralServerGroupsPoller.java

```diff
@@ -0,0 +1,249 @@
+          // server group exists -- destroy the server group (will also remove all associated tags)
+          jobs.add(buildDestroyServerGroupOperation(ephemeralServerGroupTag));
+        } else {
+          // server group no longer exists -- remove the tags
+          jobs.add(buildDeleteEntityTagsOperation(ephemeralServerGroupTag));
+        }
+
+        Map<String, Object> cleanupOperation = buildCleanupOperation(ephemeralServerGroupTag, jobs);
+        log.info((String) cleanupOperation.get("name"));
+
+        User systemUser = new User();
+        systemUser.setUsername("spinnaker");
+        systemUser.setAllowedAccounts(Collections.singletonList(ephemeralServerGroupTag.account));
+
+        AuthenticatedRequest.propagate(() -> executionLauncher.start(
+          Execution.ExecutionType.ORCHESTRATION,
+          objectMapper.writeValueAsString(cleanupOperation)
+        ), systemUser).call();
+
+        triggeredCounter.increment();
+      } catch (Exception e) {
+        log.error("Failed to destroy ephemeral server group (id: {})", ephemeralServerGroupTag.id, e);
+        errorsCounter.increment();
+      }
+    }
+  }
+
+  private List<EphemeralServerGroupTag> fetchEphemeralServerGroupTags() {
+    List<EntityTags> allEntityTags = retrySupport.retry(() -> objectMapper.convertValue(
+      oortService.getEntityTags(ImmutableMap.<String, String>builder()
+        .put("tag:" + TTL_TAG, "*")
+        .put("entityType", "servergroup")
+        .build()
+      ),
+      new TypeReference<List<EntityTags>>() {
+      }
+    ), 15, 2000, false);
+
+    return allEntityTags.stream()
+      .map(e -> e.tags.stream()
+        .filter(t -> TTL_TAG.equalsIgnoreCase(t.name))
+        .map(t -> {
+          EphemeralServerGroupTag ephemeralServerGroupTag = objectMapper.convertValue(t.value, EphemeralServerGroupTag.class);
+          ephemeralServerGroupTag.id = e.id;
+          ephemeralServerGroupTag.account = e.entityRef.account;
+          ephemeralServerGroupTag.location = e.entityRef.region;
+          ephemeralServerGroupTag.application = e.entityRef.application;
+          ephemeralServerGroupTag.serverGroup = e.entityRef.entityId;
+
+          return ephemeralServerGroupTag;
+        })
+        .findFirst()
+        .orElse(null)
+      )
+      .filter(Objects::nonNull)
+      .filter(t -> t.expiry.isBefore(ZonedDateTime.now(ZoneOffset.UTC)))
+      .collect(Collectors.toList());
+  }
+
+  private Map<String, Object> buildCleanupOperation(EphemeralServerGroupTag ephemeralServerGroupTag,
+                                                    List<Map<String, Object>> jobs) {
+    return ImmutableMap.<String, Object>builder()
+      .put("application", ephemeralServerGroupTag.application)
+      .put(
+        "name",
+        format(
+          "Destroy Ephemeral Server Group: %s",
+          ephemeralServerGroupTag.serverGroup
+        )
+      )
+      .put("user", "spinnaker")
+      .put("stages", jobs)
+      .build();
```

Comment:
I need to remember this immutable map builder... great improvement over the hashmap shenanigans.

Reply:
Great may be a stretch. It's nice, I'll remember it.